presto-main/src/test/java/io/prestosql/operator/TestGroupedTopNRankBuilder.java

```diff
@@ -0,0 +1,283 @@
+                    .row(0.1, 1)
+                    .row(0.2, 3)
+                    .row(0.2, 3)
+                    .row(0.2, 3)
+                    .build();
+            assertPageEquals(ImmutableList.of(DOUBLE, BIGINT), getOnlyElement(output), expected);
+        }
+        else {
+            Page expected = rowPageBuilder(DOUBLE)
+                    .row(0.1)
+                    .row(0.1)
+                    .row(0.2)
+                    .row(0.2)
+                    .row(0.2)
+                    .build();
+            assertPageEquals(ImmutableList.of(DOUBLE), getOnlyElement(output), expected);
+        }
+    }
+
+    @Test(dataProvider = "produceRanking")
+    public void testMultiGroupTopN(boolean produceRanking)
+    {
+        List<Type> types = ImmutableList.of(BIGINT, DOUBLE);
+
+        GroupByHash groupByHash = createGroupByHash(ImmutableList.of(types.get(0)), ImmutableList.of(0), NOOP);
+        GroupedTopNBuilder groupedTopNBuilder = new GroupedTopNRankBuilder(
+                types,
+                new SimplePageWithPositionComparator(types, ImmutableList.of(1), ImmutableList.of(ASC_NULLS_LAST), typeOperators),
+                new SimplePageWithPositionEqualsAndHash(types, ImmutableList.of(1), blockTypeOperators),
+                3,
+                produceRanking,
+                groupByHash);
+
+        // Expected effect:
+        // Group 0 [0.2 x 1 => rank=1, 0.3 x 3 => rank=2]
+        // Group 1 [0.2 x 1 => rank=1]
+        assertTrue(groupedTopNBuilder.processPage(
+                rowPageBuilder(types)
+                        .row(0L, 0.3)
+                        .row(0L, 0.3)
+                        .row(0L, 0.3)
+                        .row(0L, 0.2)
+                        .row(1L, 0.2)
+                        .build()).process());
+
+        // Page should be dropped, because all values too large to be considered
+        assertTrue(groupedTopNBuilder.processPage(
+                rowPageBuilder(types)
+                        .row(0L, 0.4)
+                        .row(1L, 0.4)
+                        .build()).process());
+
+        // Next page should cause evict 0.3 from group 0, which should cause the first page to be compacted
+        // Expected effect:
+        // Group 0 [0.1 x 1 => rank=1, 0.2 x 2 => rank=2]
+        // Group 1 [0.2 x 2 => rank=1, 0.3 x 2 => rank=3]
+        assertTrue(groupedTopNBuilder.processPage(
+                rowPageBuilder(types)
+                        .row(0L, 0.1)
+                        .row(1L, 0.2)
+                        .row(0L, 0.3)
+                        .row(0L, 0.2)
+                        .row(1L, 0.5)
+                        .row(1L, 0.4)
+                        .row(1L, 0.3)
+                        .row(1L, 0.3)
+                        .build()).process());
+
+        List<Page> output = ImmutableList.copyOf(groupedTopNBuilder.buildResult());
+        assertEquals(output.size(), 1);
+
+        if (produceRanking) {
```

Comment:
Could you extract this into separate method, e.g:
```
void assertAccumulatorOutput(GroupedTopNBuilder builder, boolean produceRanking, Page result) {
   ...
}
```
Then you could use it:
```
assertAccumulatorOutput(
  groupedTopNBuilder,
  produceRankings,
  rowPageBuilder(types)
   .row(0L, 0.1)
   .row(1L, 0.2)
   .row(0L, 0.3)
   .row(0L, 0.2)
   .row(1L, 0.5)
   .row(1L, 0.4)
   .row(1L, 0.3)
   .row(1L, 0.3)
   .build())
```

Reply:
Actually, this doesn't save anything since the code bulk comes from building the expected page results based on produceRankings, and not very much from the actual assertion itself.

Reply:
```
List<Page> expectedPages = pages.stream()
  .map(page -> page.getColumns(0))
  .collect(toImmutableList())
```